iothub/device/tests/Transport/AmqpIot/AmqpIotErrorAdapterTests.cs

```diff
@@ -0,0 +1,160 @@
+            ex.ErrorCode.Should().Be(iotHubClientErrorCode);
+        }
+
+        [TestMethod]
+        [DataRow("amqp:resource-limit-exceeded", IotHubClientErrorCode.QuotaExceeded)]
+        [DataRow("amqp:link:message-size-exceeded", IotHubClientErrorCode.MessageTooLarge)]
+        [DataRow("amqp:unauthorized-access", IotHubClientErrorCode.Unauthorized)]
+        [DataRow("amqp:not-found", IotHubClientErrorCode.DeviceNotFound)]
+        [DataRow(AmqpIotConstants.Vendor + ":timeout", IotHubClientErrorCode.Timeout)]
+        [DataRow(AmqpIotConstants.Vendor + ":message-lock-lost", IotHubClientErrorCode.DeviceMessageLockLost)]
+        [DataRow(AmqpIotConstants.Vendor + ":device-container-throttled", IotHubClientErrorCode.Throttled)]
+        [DataRow(AmqpIotConstants.Vendor + ":iot-hub-suspended", IotHubClientErrorCode.Suspended)]
+        public void AmqpIotErrorAdapter_ToIotHubClientContract_Error(string amqpErrorCode, IotHubClientErrorCode iotHubClientErrorCode)
+        {
+            var error = new Error
+            {
+                Condition = amqpErrorCode,
+            };
+            IotHubClientException ex = AmqpIotErrorAdapter.ToIotHubClientContract(error);
+            ex.ErrorCode.Should().Be(iotHubClientErrorCode);
+        }
+
+        [TestMethod]
+        [DataRow("amqp:resource-locked")]
+        [DataRow("amqp:connection:forced")]
+        [DataRow("amqp:connection:framing-error")]
+        [DataRow("amqp:connection:redirect")]
+        [DataRow("amqp:session:window-violation")]
+        [DataRow("amqp:session-errant-link")]
+        [DataRow("amqp:session:handle-in-use")]
+        [DataRow("amqp:session:unattached-handle")]
+        [DataRow("amqp:link:detach-forced")]
+        [DataRow("amqp:link:transfer-limit-exceeded")]
+        [DataRow("amqp:link:redirect")]
+        [DataRow("amqp:link:stolen")]
+        public void AmqpIotErrorAdapter_ToIotHubClientContract_IsTransient(string amqpErrorCode)
+        {
+            var error = new Error
+            {
+                Condition = amqpErrorCode,
+            };
+            IotHubClientException ex = AmqpIotErrorAdapter.ToIotHubClientContract(new AmqpException(error));
+            ex.IsTransient.Should().BeTrue();
+        }
+
+        [TestMethod]
+        public void AmqpIotErrorAdapter_ToIotHubClientContract_Error_Null()
+        {
+            IotHubClientException ex = AmqpIotErrorAdapter.ToIotHubClientContract((Error) null);
+            ex.Should().BeEquivalentTo(new IotHubClientException("Unknown error."));
+        }
+
+        [TestMethod]
+        public void AmqpIotExceptionAdapter_ConvertToIotHubException_InvalidOperationException()
+        {
+            InvalidOperationException exception = new("message");
+            using var socket = new ClientWebSocket();
+            using var source = new ClientWebSocketTransport(socket, null, null);
+            source.SafeClose();
```

Comment:
Is the SafeClose required here? Would mocking a websocket or passing a null socket instance instead work?

Reply:
SafeClose is required as the specific condition in the code was not covered by a unit test. mocking a websocket does not work as its a sealed class.